namer/namer.cc

```diff
@@ -1380,50 +1380,62 @@ class SymbolDefiner {
+        // Note: this loop is accidentally quadratic. We run deleteFieldViaFullNameHash once per field
+        // previously defined in this file, then in each call look at each member of that field's owner.
         for (const auto &[memberName, memberSym] : owner.data(ctx)->members()) {
-            if (!memberSym.isMethod()) {
+            if (memberSym.isClassOrModule()) {
+                // Safeguard against a collision function in our `FullNameHash` function (e.g., what
+                // if `foo()` and `::Foo` have the same `FullNameHash`, but were given two different
+                // `NameRef` IDs and thus don't collide in the `members` list?).
+                //
+                // Any other collisions are fine, because we will already be re-entering the new
+                // definitions later on in incremental namer. It's only classes that we definitely
+                // never want to delete accidentally.
                 continue;
             }
-            auto memberMethod = memberSym.asMethodRef();
-
-            auto memberNameToHash = memberName;
-            if (memberNameToHash.kind() == core::NameKind::UNIQUE) {
-                auto &uniqueData = memberNameToHash.dataUnique(ctx);
-                if (uniqueData->uniqueNameKind == core::UniqueNameKind::MangleRename ||
-                    uniqueData->uniqueNameKind == core::UniqueNameKind::MangleRenameOverload ||
-                    uniqueData->uniqueNameKind == core::UniqueNameKind::Overload) {
-                    memberNameToHash = uniqueData->original;
-                }
-            }
 
-            auto memberFullNameHash = core::FullNameHash(ctx, memberNameToHash);
-            if (memberFullNameHash != oldDefHash.nameHash) {
+            if (!matchesFullNameHash(ctx, memberName, oldNameHash)) {
                 continue;
             }
 
-            toDelete.emplace_back(memberMethod);
+            toDelete.emplace_back(memberSym);
         }
 
-        for (auto oldMethod : toDelete) {
-            oldMethod.data(ctx)->removeLocsForFile(ctx.file);
-            if (oldMethod.data(ctx)->locs().empty()) {
-                ctx.state.deleteMethodSymbol(oldMethod);
+        for (auto oldSymbol : toDelete) {
+            oldSymbol.removeLocsForFile(ctx, ctx.file);
+            if (oldSymbol.locs(ctx).empty()) {
+                switch (oldSymbol.kind()) {
+                    case core::SymbolRef::Kind::Method:
+                        ctx.state.deleteMethodSymbol(oldSymbol.asMethodRef());
+                        break;
+                    case core::SymbolRef::Kind::FieldOrStaticField:
+                        ctx.state.deleteFieldSymbol(oldSymbol.asFieldRef());
+                        break;
+                    case core::SymbolRef::Kind::TypeMember:
+                    case core::SymbolRef::Kind::ClassOrModule:
+                    case core::SymbolRef::Kind::TypeArgument:
+                        ENFORCE(false);
```

Comment:
Does this need a `break;` or something after it to not warn in non-debug builds?